gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -493,4 +531,31 @@ public boolean apply(KafkaTopic input) {
       return this.topicNamePattern.matcher(input.getName()).matches();
     }
   }
+  
+  private class WorkUnitCreator implements Runnable{
+
+    private final KafkaSource<S,D> kafkaSource;
+    private final KafkaTopic topic;
+    private final SourceState state;
+    private final Optional <State> topicSpecificState;
+    private final Map<String, List<WorkUnit>> allTopicWorkUnits;
+    
+    WorkUnitCreator(KafkaTopic topic, SourceState state, Optional<State> topicSpecificState, 
+        KafkaSource<S,D> kafkaSource, Map<String, List<WorkUnit>> workUnits){
+      this.kafkaSource = kafkaSource;
+      this.topic = topic;
+      this.state = state;
+      this.topicSpecificState = topicSpecificState;
+      this.allTopicWorkUnits = workUnits;
+    }
+    
+    @Override
+    public void run() {
+      List<WorkUnit> oneTopicWorkUnits = 
+          this.kafkaSource.getWorkUnitsForTopic(this.topic, this.state, topicSpecificState);
```

Comment:
You don't need to pass `kafkaSource` to this class. Here you can do `KafkaSource.this.getWorkUnitsForTopic`